v2/devdoc/logging.md

```diff
@@ -0,0 +1,221 @@
+# `c-logging` requirements
+
+## Basic functionality
+
+`c-logging` shall expose APIs allowing the user to log events that can be used for observing the software activity, debugging, etc.
+
+The exposed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    logger_log(log_level, log_context, __FILE__, __FUNCTION__, __LINE__, format MU_IFCOMMALOGIC(MU_COUNT_ARG(__VA_ARGS__)) __VA_ARGS__)
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify properties for a context, they can use LOG_CONTEXT_PROPERTY
+// LOG_CONTEXT_PROPERTY(property_type, property_name, property_value)
+
+// when a user wants to specify string fields in a context using printf-style formatting, they can use LOG_CONTEXT_STRING_PROPERTY
+// LOG_CONTEXT_STRING_PROPERTY(property_name, property_conversion_specifier, property_value)
+```
+
+Example usage:
+
+```c
+    // chained allocated context
+    LOG_CONTEXT_HANDLE context_1;
+    LOG_CONTEXT_CREATE(context_1, NULL, LOG_CONTEXT_STRING_PROPERTY(name, "%s", "haga"));
+    LOG_CONTEXT_HANDLE context_2;
+    LOG_CONTEXT_CREATE(context_2, context_1, LOG_CONTEXT_STRING_PROPERTY(last_name, "%s", "uaga"), LOG_CONTEXT_PROPERTY(int32_t, age, 42));
+
+    LOGGER_LOG(LOG_LEVEL_CRITICAL, context_1, "log context 1 data, value of some_var = %d", 42);
+    LOGGER_LOG(LOG_LEVEL_VERBOSE, context_2, "log context 2 data, some other string value is %s", "mumu");
+
+    LOG_CONTEXT_DESTROY(context_1);
+
+    // stack allocated context
+    LOG_CONTEXT_LOCAL_DEFINE(local_context, context_2, LOG_CONTEXT_STRING_PROPERTY(the_knights_that_say, "%s", "Nee!"));
+
+    LOGGER_LOG(LOG_LEVEL_VERBOSE, &local_context, "log with local context");
+
+    LOG_CONTEXT_DESTROY(context_2);
+```
+
+The above example produces:
+
+```
+LOG_LEVEL_CRITICAL Time:Thu Aug 11 12:51:55 2022 File:G:\w\c-logging\v2\tests\logger_int\main.c Func:main Line:96 name=haga last name=uaga age=42  log context 1 data, value of some_var = 42
+LOG_LEVEL_VERBOSE Time:Thu Aug 11 12:51:55 2022 File:G:\w\c-logging\v2\tests\logger_int\main.c Func:main Line:97 name=haga last name=uaga age=42  log context 2 data, some other string value is mumu
```

Comment:
>name=haga last name=uaga age=42 [](http://example.com/codeflow?start=113&length=31)

Would be nice if we could format (or chose how to format) these "context" property bags, maybe as JSON/XML or some other structured data, because at the moment they are completely unstructured (which is also fine). Depends on who reads them.

Reply:
This sounds like a feature for the tasks that never get done.

Reply:
Let's consider this once we have the v2 (when time comes)